agent/agent.go

```diff
@@ -346,44 +350,54 @@ func (a *Agent) Run(ctx context.Context) error {
 }
 
 // updateStructValuesHelper updates the reflect config, returning the original plugin otherwise
-func updateStructValuesHelper(pluginPtr reflect.Value, newConfig map[string]interface{}) (reflect.Value, error) {
-
+func updateStructValuesHelper(pluginPtr reflect.Value, newConfig map[string]interface{}) (reflect.Value, map[string]interface{}, error) {
 	plugin := pluginPtr.Elem() // extract Value of type interface{} from Value pointer to interface
+	tomlMap := map[string]interface{}{}
+	pType := plugin.Type()
 
 	if plugin.Kind() == reflect.Struct {
 
 		// initial check for errors
 		for configKey, configValue := range newConfig {
 			pluginField := plugin.FieldByName(configKey) // cast new value as Value
 			reflectedNew := reflect.ValueOf(configValue)
+
 			// if any errors found, return original plugin
 			if !(pluginField.IsValid()) {
-				return pluginPtr.Elem(), fmt.Errorf("invalid field name %s", configKey)
+				return pluginPtr.Elem(), map[string]interface{}{}, fmt.Errorf("invalid field name %s", configKey)
 			} else if !(pluginField.CanSet()) {
-				return pluginPtr.Elem(), fmt.Errorf("unsettable field %s", configKey)
+				return pluginPtr.Elem(), map[string]interface{}{}, fmt.Errorf("unsettable field %s", configKey)
 			} else if !(pluginField.Type() == reflectedNew.Type()) {
-				return pluginPtr.Elem(), fmt.Errorf("value type mismatch for field %s", configKey)
+				return pluginPtr.Elem(), map[string]interface{}{}, fmt.Errorf("value type mismatch for field %s", configKey)
 			}
-		}
 
-		// if no error, update all
-		for configKey, configValue := range newConfig {
-			pluginField := plugin.FieldByName(configKey) // cast new value as Value
-			reflectedNew := reflect.ValueOf(configValue)
-			pluginField.Set(reflectedNew)
+			// if no error, update all
+			for configKey, configValue := range newConfig {
```

Comment:
Should this be outside of the "initial check for errors" for loop?

Reply:
Why are we looping over all these KV Pairs twice?

Reply:
Shouldn't this inner loop be merged into the outer loop, with the redundant stuff removed?

Reply:
ah shoot I think this got fcked by the rebase / merge conflicts. We don't want the loops to be nested but we do want separate ones. The reason is that we want all of the fields to be valid before making any updates. Otherwise if any of the fields are incorrect we can catch them in the first loop and return early.